client-apis/microprofile-restclient/src/main/java/org/wildfly/swarm/microprofile/restclient/InvocationContextImpl.java

```diff
@@ -0,0 +1,168 @@
+    InvocationContextImpl(Object target, Method method, Object[] args, List<InterceptorInvocation> chain) {
+        this.target = target;
+        this.method = method;
+        this.args = args;
+        this.contextData = new HashMap<>();
+        this.position = 0;
+        this.chain = chain;
+    }
+
+    boolean hasNextInterceptor() {
+        return position < chain.size();
+    }
+
+    protected Object invokeNext() throws Exception {
+        int oldPosition = position;
+        try {
+            return chain.get(position++).invoke(this);
+        } finally {
+            position = oldPosition;
+        }
+    }
+
+    protected Object interceptorChainCompleted() throws Exception {
+        try {
+            return method.invoke(target, args);
+        } catch (InvocationTargetException e) {
+            if (e.getCause() instanceof ResponseProcessingException) {
+                ResponseProcessingException rpe = (ResponseProcessingException) e.getCause();
+                // Note that the default client engine leverages a single connection
+                // MP FT: we need to close the response otherwise we would not be able to retry if the method returns javax.ws.rs.core.Response
+                rpe.getResponse().close();
+                Throwable cause = rpe.getCause();
+                if (cause instanceof RuntimeException) {
+                    throw (RuntimeException) cause;
+                }
+            }
+            throw e;
+        }
+    }
+
+    @Override
+    public Object proceed() throws Exception {
+        try {
+            if (hasNextInterceptor()) {
+                return invokeNext();
+            } else {
+                return interceptorChainCompleted();
+            }
+        } catch (InvocationTargetException e) {
+            Throwable cause = e.getCause();
+            if (cause instanceof Error) {
+                throw (Error) cause;
+            }
+            if (cause instanceof Exception) {
+                throw (Exception) cause;
+            }
+            throw new RuntimeException(cause);
+        }
+    }
+
+    @Override
+    public Object getTarget() {
+        return target;
+    }
+
+    @Override
+    public Method getMethod() {
+        return method;
+    }
+
+    @Override
+    public Constructor<?> getConstructor() {
+        return null;
+    }
+
+    @Override
+    public Object[] getParameters() throws IllegalStateException {
+        return args;
+    }
+
+    @Override
+    public void setParameters(Object[] params) throws IllegalStateException, IllegalArgumentException {
+        this.args = params;
+    }
+
+    @Override
+    public Map<String, Object> getContextData() {
+        return contextData;
+    }
+
+    @Override
+    public Object getTimer() {
+        return null;
+    }
+
+    static class InterceptorInvocation {
+
+        @SuppressWarnings("rawtypes")
+        private final Interceptor interceptor;
+
+        private final Object interceptorInstance;
```

Comment:
Is this an _intercepted_ instance?

Reply:
No. It's really an *interceptor* instance (e.g. instance of [HystrixCommandInterceptor](https://github.com/wildfly-swarm/wildfly-swarm/blob/master/fractions/microprofile/microprofile-fault-tolerance/src/main/java/org/wildfly/swarm/microprofile/faulttolerance/deployment/HystrixCommandInterceptor.java)). Whereas the `Interceptor` represents interceptor metadata.